File: fortran/src/token.f90
! Token expansion for wildling patterns.
module wl_token
  use iso_fortran_env, only: int64
  use wl_util
  implicit none
  private

  public :: token_t, token_init, token_free, token_count, token_get, int_pow

  type :: token_t
    character(len=:), allocatable :: src
    integer :: start_length = 1
    integer :: end_length = 1
    type(str_list) :: variants
    integer(int64) :: count = 0
  end type token_t

contains

  pure integer(int64) function int_pow(base, exp) result(r)
    integer(int64), intent(in) :: base
    integer, intent(in) :: exp
    integer :: i
    r = 1_int64
    do i = 1, exp
      r = r * base
    end do
  end function int_pow

  subroutine token_init(tok, src, start_length, end_length, variants)
    type(token_t), intent(out) :: tok
    character(len=*), intent(in) :: src
    integer, intent(in) :: start_length, end_length
    type(str_list), intent(inout) :: variants
    integer :: length_val
    integer(int64) :: nvar
    tok%src = src
    if (start_length >= 0) then
      tok%start_length = start_length
    else
      tok%start_length = 1
    end if
    if (end_length >= 0) then
      tok%end_length = end_length
    else
      tok%end_length = 1
    end if
    tok%variants = variants
    call str_list_init(variants)
    tok%count = 0_int64
    nvar = int(tok%variants%n, int64)
    do length_val = tok%start_length, tok%end_length
      tok%count = tok%count + int_pow(nvar, length_val)
    end do
  end subroutine token_init

  subroutine token_free(tok)
    type(token_t), intent(inout) :: tok
    if (allocated(tok%src)) deallocate(tok%src)
    call str_list_free(tok%variants)
    tok%count = 0
  end subroutine token_free

  pure integer(int64) function token_count(tok) result(n)
    type(token_t), intent(in) :: tok
    n = tok%count
  end function token_count

  function token_get(tok, index) result(out)
    type(token_t), intent(in) :: tok
    integer(int64), intent(in) :: index
    character(len=:), allocatable :: out
    integer(int64) :: index_with_offset, offset_count, nvar
    integer :: string_length, variant_index, i
    out = ''
    if (index > tok%count - 1_int64 .or. index < 0_int64) return
    if (index == 0_int64 .and. tok%start_length == 0) return

    index_with_offset = index
    string_length = tok%start_length
    nvar = int(tok%variants%n, int64)
    do string_length = tok%start_length, tok%end_length
      offset_count = int_pow(nvar, string_length)
      if (index_with_offset < offset_count) exit
      index_with_offset = index_with_offset - offset_count
    end do

    out = ''
    do i = 1, string_length
      if (tok%variants%n == 0) exit
      variant_index = int(mod(index_with_offset, nvar)) + 1
      index_with_offset = index_with_offset / nvar
      out = out // tok%variants%items(variant_index)%s
    end do
  end function token_get

end module wl_token

File: pascal/src/token.pas
unit Token;

{$mode objfpc}{$H+}

interface

type
  TToken = class
  private
    FSrc: string;
    FStartLength: Integer;
    FEndLength: Integer;
    FVariants: array of string;
    FCount: Int64;
  public
    constructor Create(const ASrc: string; AStartLength, AEndLength: Integer;
      const AVariants: array of string);
    function Count: Int64;
    function Src: string;
    function Get(Index: Int64): string;
  end;

function IntPow(Base: Int64; Exp: Integer): Int64;
function CreateToken(const ASrc: string; AStartLength, AEndLength: Integer;
  const AVariants: array of string): TToken;
function CreateTokenDefaults(const ASrc: string; const AVariants: array of string): TToken;

implementation

function IntPow(Base: Int64; Exp: Integer): Int64;
var
  I: Integer;
begin
  Result := 1;
  for I := 0 to Exp - 1 do
    Result := Result * Base;
end;

constructor TToken.Create(const ASrc: string; AStartLength, AEndLength: Integer;
  const AVariants: array of string);
var
  LengthVal, I: Integer;
begin
  inherited Create;
  FSrc := ASrc;
  if AStartLength >= 0 then
    FStartLength := AStartLength
  else
    FStartLength := 1;
  if AEndLength >= 0 then
    FEndLength := AEndLength
  else
    FEndLength := 1;
  SetLength(FVariants, Length(AVariants));
  for I := 0 to High(AVariants) do
    FVariants[I] := AVariants[I];
  FCount := 0;
  for LengthVal := FStartLength to FEndLength do
    FCount := FCount + IntPow(Length(FVariants), LengthVal);
end;

function TToken.Count: Int64;
begin
  Result := FCount;
end;

function TToken.Src: string;
begin
  Result := FSrc;
end;

function TToken.Get(Index: Int64): string;
var
  IndexWithOffset: Int64;
  StringLength: Integer;
  OffsetCount: Int64;
  VariantIndex: Integer;
  Parts: string;
  I: Integer;
  VariantCount: Integer;
begin
  if (Index > FCount - 1) or (Index < 0) then
    Exit('');

  if (Index = 0) and (FStartLength = 0) then
    Exit('');

  IndexWithOffset := Index;
  StringLength := FStartLength;
  for StringLength := FStartLength to FEndLength do
  begin
    OffsetCount := IntPow(Length(FVariants), StringLength);
    if IndexWithOffset < OffsetCount then
      Break;
    IndexWithOffset := IndexWithOffset - OffsetCount;
  end;

  Parts := '';
  VariantCount := Length(FVariants);
  for I := 1 to StringLength do
  begin
    if VariantCount = 0 then
      Break;
    VariantIndex := IndexWithOffset mod VariantCount;
    IndexWithOffset := IndexWithOffset div VariantCount;
    Parts := Parts + FVariants[VariantIndex];
  end;
  Result := Parts;
end;

function CreateToken(const ASrc: string; AStartLength, AEndLength: Integer;
  const AVariants: array of string): TToken;
begin
  Result := TToken.Create(ASrc, AStartLength, AEndLength, AVariants);
end;

function CreateTokenDefaults(const ASrc: string; const AVariants: array of string): TToken;
begin
  Result := TToken.Create(ASrc, 1, 1, AVariants);
end;

end.
